api/testing/restapi_test.go

```diff
@@ -0,0 +1,686 @@
+	id := "myid"
+
+	gomock.InOrder(
+		ts.MockDriver().
+			EXPECT().
+			Delete(id).
+			Return(nil).
+			Times(1),
+	)
+
+	driverclient := volumeclient.VolumeDriver(ts.client)
+
+	err = driverclient.Delete(id)
+	assert.Nil(t, err)
+}
+
+func TestVolumeDeleteFailed(t *testing.T) {
+
+	ts := setupMocks()
+
+	defer ts.Stop()
+
+	var err error
+	ts.client, err = volumeclient.NewDriverClient(host, driver, version, "")
+
+	// Setup mock
+
+	id := "myid"
+
+	gomock.InOrder(
+		ts.MockDriver().
+			EXPECT().
+			Delete(gomock.Any()).
+			Return(fmt.Errorf("error in delete")).
+			Times(1),
+	)
+
+	driverclient := volumeclient.VolumeDriver(ts.client)
+
+	err = driverclient.Delete(id)
+	assert.NotNil(t, err)
+}
+
+func TestSnapshotCreateSuccess(t *testing.T) {
+
+	ts := setupMocks()
+
+	defer ts.Stop()
+
+	id := "myid"
+	name := "snapName"
+
+	var err error
+	ts.client, err = volumeclient.NewDriverClient(host, driver, version, "")
+
+	assert.Nil(t, err)
+
+	req := &api.SnapCreateRequest{Id: id,
+		Locator:  &api.VolumeLocator{Name: name},
+		Readonly: true,
+	}
+
+	//mock Snapshot call
+	gomock.InOrder(
+		ts.MockDriver().
+			EXPECT().
+			Snapshot(gomock.Any(), gomock.Any(), gomock.Any()).
+			Return(id, nil).
+			Times(1),
+	)
+
+	//make the call
+	driverclient := volumeclient.VolumeDriver(ts.client)
+
+	res, err := driverclient.Snapshot(req.GetId(), req.GetReadonly(), req.GetLocator())
+
+	assert.Nil(t, err)
+	assert.Equal(t, id, res)
+
+}
+
+func TestVolumeInspectSuccess(t *testing.T) {
+
+	ts := setupMocks()
+
+	defer ts.Stop()
+
+	var err error
+	ts.client, err = volumeclient.NewDriverClient(host, driver, version, "")
+
+	assert.Nil(t, err)
+
+	id := "myid"
+	var size uint64 = 1234
+	name := "inspectVol"
+
+	gomock.InOrder(
+		ts.MockDriver().
+			EXPECT().
+			Inspect([]string{id}).
+			Return([]*api.Volume{
+				&api.Volume{
+					Id: id,
+					Locator: &api.VolumeLocator{
+						Name: name,
+					},
+					Spec: &api.VolumeSpec{
+						Size: size,
+					},
+				},
+			}, nil).
+			Times(1),
+	)
+
+	driverclient := volumeclient.VolumeDriver(ts.client)
+
+	res, err := driverclient.Inspect([]string{id})
+
+	assert.Nil(t, err)
+	assert.NotNil(t, res)
+	assert.NotEmpty(t, res)
+	assert.Equal(t, res[0].GetId(), id)
+
+}
+
+func TestVolumeSetSuccess(t *testing.T) {
+
+	ts := setupMocks()
+
+	defer ts.Stop()
+
+	var err error
+	ts.client, err = volumeclient.NewDriverClient(host, driver, version, "")
+
+	assert.Nil(t, err)
+
+	// create a volume request
+
+	name := "myvol"
+	id := "myid"
+	size := uint64(10)
+
+	req := &api.VolumeSetRequest{
+		Options: map[string]string{},
+		Action: &api.VolumeStateAction{
+			Attach: api.VolumeActionParam_VOLUME_ACTION_PARAM_ON,
+			Mount:  api.VolumeActionParam_VOLUME_ACTION_PARAM_ON},
+		Locator: &api.VolumeLocator{Name: name},
+		Spec:    &api.VolumeSpec{Size: size},
+	}
+
+	gomock.InOrder(
+		ts.MockDriver().
+			EXPECT().
+			Set(id, req.GetLocator(), req.GetSpec()).
+			Return(nil),
+
+		ts.MockDriver().
```

Comment:
If I am not wrong wouldn't the Set() call with this request object result into an Attach and Mount to be called as well ?

Reply:
Ideally it should but Set method takes only the VolumeLocator and Spec , which is what is being passed to the server by the client wrapper method when making the http rest call. 
VolumeSetRequest.VolumeStateAction is not passed by the client wrapper hence processed as nil by server.